### PR TITLE
[rest] add active dataset API

### PR DESCRIPTION
**Pull request by agners**
Support getting and setting the active dataset via REST. The active data set is returned as operational dataset encoded as TLV hex string. The `PUT` method along with the body with the dataset encoded in the same format can be used to set the current active dataset.

### CODE REVIEW

File: src/rest/json.hpp [hunk 1]
```diff
@@ -70,6 +70,18 @@ std::string Number2JsonString(const uint32_t &aNumber);
...
+ * This method parses a hex string as byte array.
+ *
+ * @param[in] hexString  String of bytes in hex
+ * @param[in] aBytes     Byte array to write to. Must be at least aMaxLength.
+ * @param[in] aMaxLength Maximum length to parse (in bytes).
+ *
+ * @returns Number of bytes effectively parsed.
+ *
+ */
+int Hex2BytesJsonString(const std::string &hexString, uint8_t *aBytes, uint8_t aMaxLength);
```
Comment (jwhui) [inline src/rest/json.hpp:83]:
```suggestion
int Hex2BytesJsonString(const std::string &aHexString, uint8_t *aBytes, uint8_t aMaxLength);
```

File: src/rest/json.hpp [hunk 2]
```diff
@@ -70,6 +70,18 @@ std::string Number2JsonString(const uint32_t &aNumber);
  */
 std::string Bytes2HexJsonString(const uint8_t *aBytes, uint8_t aLength);
 
+/**
+ * This method parses a hex string as byte array.
+ *
+ * @param[in] hexString  String of bytes in hex
+ * @param[in] aBytes     Byte array to write to. Must be at least aMaxLength.
+ * @param[in] aMaxLength Maximum length to parse (in bytes).
```
Comment (jwhui) [inline src/rest/json.hpp:78]:
```suggestion
 * @param[in] aHexString  String of bytes in hex.
 * @param[in] aBytes      Byte array to write to. Must be at least @p aMaxLength.
 * @param[in] aMaxLength  Maximum length to parse (in bytes).
```

File: src/rest/resource.cpp [hunk 3]
```diff
@@ -489,6 +499,75 @@ void Resource::Rloc(const Request &aRequest, Response &aResponse) const
...
+    otOperationalDatasetTlvs datasetTlvs;
+    otError                  error = OT_ERROR_NONE;
+    std::string    errorCode;
+
+    ret = Json::Hex2BytesJsonString(aRequest.GetBody(), datasetTlvs.mTlvs, OT_OPERATIONAL_DATASET_MAX_LENGTH);
+    if (ret < 0)
+    {
+        errorCode = GetHttpStatus(HttpStatusCode::kStatusBadRequest);
+        aResponse.SetResponsCode(errorCode);
+        return;
```
Comment (jwhui) [inline src/rest/resource.cpp:537]:
```suggestion
        ExitNow();
```

File: src/rest/resource.cpp [hunk 4]
```diff
@@ -489,6 +499,75 @@ void Resource::Rloc(const Request &aRequest, Response &aResponse) const
...
+    }
+    datasetTlvs.mLength = ret;
+
+    SuccessOrExit(otDatasetSetActiveTlvs(mInstance, &datasetTlvs));
```
Comment (jwhui) [inline src/rest/resource.cpp:541]:
```suggestion
    SuccessOrExit(error = otDatasetSetActiveTlvs(mInstance, &datasetTlvs));
```

File: src/rest/resource.cpp [hunk 5]
```diff
@@ -489,6 +499,75 @@ void Resource::Rloc(const Request &aRequest, Response &aResponse) const
...
 }
 
+void Resource::GetActiveDatasetTlvs(Response &aResponse) const
+{
+    otOperationalDatasetTlvs datasetTlvs;
+    otError                  error        = OT_ERROR_NONE;
+    std::string    body;
+    std::string    errorCode;
+
+    SuccessOrExit(otDatasetGetActiveTlvs(mInstance, &datasetTlvs));
```
Comment (jwhui) [inline src/rest/resource.cpp:509]:
```suggestion
    SuccessOrExit(error = otDatasetGetActiveTlvs(mInstance, &datasetTlvs));
```

File: src/rest/resource.cpp [hunk 6]
```diff
@@ -489,6 +499,75 @@ void Resource::Rloc(const Request &aRequest, Response &aResponse) const
...
+        otbrLogWarning("Failed to set active dataset: %s", otThreadErrorToString(error));
+        ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
+    }
+}
+
+void Resource::ActiveDatasetTlvs(const Request &aRequest, Response &aResponse) const
+{
+    std::string errorCode;
+
+    if (aRequest.GetMethod() == HttpMethod::kGet)
```
Comment (jwhui) [inline src/rest/resource.cpp:557]: nit: use switch statement?

File: src/rest/json.cpp [hunk 7]
```diff
@@ -372,6 +372,11 @@ std::string Bytes2HexJsonString(const uint8_t *aBytes, uint8_t aLength)
     return ret;
 }
 
+int Hex2BytesJsonString(const std::string &hexString, uint8_t *aBytes, uint8_t aMaxLength)
```
Comment (jwhui) [inline src/rest/json.cpp:375]:
```suggestion
int Hex2BytesJsonString(const std::string &aHexString, uint8_t *aBytes, uint8_t aMaxLength)
```

File: src/rest/json.cpp [hunk 8]
```diff
@@ -372,6 +372,11 @@ std::string Bytes2HexJsonString(const uint8_t *aBytes, uint8_t aLength)
...
+{
+    return otbr::Utils::Hex2Bytes(hexString.c_str(), aBytes, aMaxLength);
```
Comment (jwhui) [inline src/rest/json.cpp:377]:
```suggestion
    return otbr::Utils::Hex2Bytes(aHexString.c_str(), aBytes, aMaxLength);
```

**Review by jwhui — COMMENTED** [review]
Note that pretty check is failing. You can run `./script/make-pretty` to automatically format.

File: src/rest/resource.cpp (hunk 3 above)
Comment (agners) [reply src/rest/resource.cpp:537]: This would overwrite the HTTP return code...

Comment (jwhui) [reply src/rest/resource.cpp:537]: The entire `exit:` block is conditional on `error != OT_ERROR_NONE`. What am I missing?

Comment (agners) [reply src/rest/resource.cpp:537]: Right!

**Review by jwhui — APPROVED** [review]

**Comment by jwhui** [comment]
Do you mind if I squash and merge these commits? Or would you prefer to keep these commits separate? If you want to keep them separate, can you include the PR # in the commit title? For example:
- `[rest] add active dataset support (#1658)`

**Comment by agners** [comment]
Squashing them is fine by me.

**Pull request merged**